modules/data.remote/inst/satellitetools/biophys_xarray.py

```diff
@@ -0,0 +1,408 @@
+    pixel_vector = np.zeros(shape=(11, len(single_date_dict[band_list[0]])))
+
+    for i, b in enumerate(band_list):
+        pixel_vector[i, :] = np.array(single_date_dict[b]) / 10000.0
+
+    pixel_vector[8, :] = np.cos(np.radians(single_date_dict["view_zenith"]))
+    pixel_vector[9, :] = np.cos(np.radians(single_date_dict["sun_zenith"]))
+    pixel_vector[10, :] = np.cos(
+        np.radians(single_date_dict["sun_azimuth"] - single_date_dict["view_azimuth"])
+    )
+    return pixel_vector
+
+
+def compute_ci_red_edge(dataset):
+    """Compute CI_Red_Edge vegetation index.
+
+    Parameters
+    ----------
+    dataset : xarray dataset
+
+    Returns
+    -------
+    xarray dataset
+        Adds 'ci_red_edge' xr array to xr dataset.
+
+    """
+    b5 = dataset.band_data.sel(band="B5")
+    b7 = dataset.band_data.sel(band="B7")
+    ci_red_edge = (b7 / b5) - 1
+    return dataset.assign({"ci_red_edge": ci_red_edge})
+
+
+def run_snap_biophys(dataset, variable):
+    """Compute specified variable using the SNAP algorithm.
+
+    See ATBD at https://step.esa.int/docs/extra/ATBD_S2ToolBox_L2B_V1.1.pdf
+
+    Parameters
+    ----------
+    dataset : xr dataset
+        xarray dataset.
+    variable : str
+        Options 'FAPAR', 'FCOVER', 'LAI', 'LAI_Cab' or 'LAI_Cw'
+
+    Returns
+    -------
+    xarray dataset
+        Adds the specified variable array to dataset (variable name in
+        lowercase).
+
+    """
+    # generate view angle bands/layers
+    vz = (
+        np.ones_like(dataset.band_data[:, 0, :, :]).T
+        * np.cos(np.radians(dataset.view_zenith)).values
+    )
+    vz = vz[..., np.newaxis]
+    vzarr = xr.DataArray(
+        vz,
+        coords=[dataset.y, dataset.x, dataset.time, ["view_zenith"]],
+        dims=["y", "x", "time", "band"],
+    )
+
+    sz = (
+        np.ones_like(dataset.band_data[:, 0, :, :]).T
+        * np.cos(np.radians(dataset.sun_zenith)).values
+    )
+    sz = sz[..., np.newaxis]
+    szarr = xr.DataArray(
+        sz,
+        coords=[dataset.y, dataset.x, dataset.time, ["sun_zenith"]],
+        dims=["y", "x", "time", "band"],
+    )
+
+    raz = (
+        np.ones_like(dataset.band_data[:, 0, :, :]).T
+        * np.cos(np.radians(dataset.sun_azimuth - dataset.view_azimuth)).values
+    )
+    raz = raz[..., np.newaxis]
+    razarr = xr.DataArray(
+        raz,
+        coords=[dataset.y, dataset.x, dataset.time, ["relative_azimuth"]],
+        dims=["y", "x", "time", "band"],
+    )
+
+    newarr = xr.concat([dataset.band_data, vzarr, szarr, razarr], dim="band")
+    newarr = newarr.stack(xy=("x", "y"))
+    arr = xr.apply_ufunc(
+        _compute_variable,
+        newarr,
+        input_core_dims=[["band", "xy"]],
+        output_core_dims=[["xy"]],
+        kwargs={"variable": variable},
+        vectorize=True,
+    ).unstack()
+    return dataset.assign({variable.lower(): arr})
+
+
+def estimate_gpp_vi_lue(vi, daily_par, model_name="ci_red_edge_1"):
```

Comment:
This function takes us a step past the estimation of LAI and adds a (overly) simplistic GPP calculation with a lot of hard-coded parameters. @istfer is this really needed for the application you envisioned? I'd really prefer to drop this and `GPP_LUE_models` as likely to cause more trouble than benifit (e.g. attempts to calibrate or benchmark models against models).

Reply:
I think these are creeping in from Olli's code without being intended, GPP calculation is not a part of the GEE-GSOC work (it's an exercise from another project of Olli's)

I'll only be able to review this code tomorrow @ayushprd then we can comb out what we need

Reply:
The reason why I did not remove this plus the other features such as calculation of FAPAR, FCOVER, etc is that I thought these functions could be helpful to you all later on even if not used immediately. The other file included in this PR, `s2lai.py` simply calculates the LAI and does not estimate GPP or make use of the other functions present in SNAP. However, I understand these are not helpful, so I've marked the parts based on my understanding and testing which can be removed and will not affect the calculation of LAI. Sorry for the confusion. Henceforth, I will only think of NDVI and LAI while developing this project.

Reply:
Hi Ayush, no please don't only think of NDVI and LAI while developing this project, we will need all sorts of data types and derivations. But as Sentinel2-GEE-NDVI-LAI pipeline already exists, that is what you can focus first of course, if that's what you mean.

It is fine that you weren't sure if other features were wanted or not. We normally want this module to give us data to compare to model predictions, but remote sensing data usually need some operation first. As you can see LAI is already more of a modeled-output than data, but most models do predict LAI, so we don't need the extra massaging it into GPP etc. Some models can already predict spectral bands, then we can directly use those and don't even need the LAI calculation step.

Thanks for marking the parts that are not needed for LAI calculation already, I'll get back to this PR/you after my morning meetings.

Reply:
I second @istfer point -- my request to remove GPP was not a blanket request to remove everything that's not LAI. I agree that FAPAR and FCOVER are useful.

As a broader point, when it comes to remote sensing there's definitely a grey area when it comes to the question of what is "data" and what is a prediction from a model that's using remote sensing as one of its inputs. Raw bands are definitely in the first category, but very few of our ecosystem models predict that so they're not particularly useful validation. Also in this category are simple mathematical transforms of raw bands (e.g. NDVI) but process-models are likewise not routinely calculating this. GPP is definitely in the second category -- satellites are by no means observing this, but instead there are models that take satellite LAI as an input (along with light, temperature, humidity, etc. as well as a bunch of vegetation specific parameters) to try and predict GPP. In between these extremes of NDVI and GPP are a lot of algorithms that fall in a grey area of being more than an analytical transform but not fully a model, with fAPAR and LAI being great examples -- they are mostly physics, but require some assumptions about the canopy structure and leaf chemistry -- and different assumptions can lead to different numbers (though generally different algorithms are all pretty close).